### PR TITLE
Enable session multiplexing

### PR DESCRIPTION
**Pull request by tgoyne**
And add an option to disable it. We previously didn't expose any SyncClientConfig settings, so most of the changes here are actually just refactoring things to expose them on RLMAppConfiguration.

This also fixes an incidental bug where mutating a RLMAppConfiguration after using it to create a RLMApp could potentially cause problems as we didn't copy the input object. This was extraordinarily unlikely to ever have been an actual problem in practice.

### CODE REVIEW

**Comment by tgoyne** [comment]
This appears to break async open timeouts.

**Comment by tgoyne** [comment]
It appears that the problem is that with session multiplexing enabled we never actually disconnect from the server, and so cannot test reconnecting with the same user.

**Review by jsflax — APPROVED** [review]

File: Realm/RLMApp_Private.h [hunk 1]
```diff
@@ -36,15 +36,15 @@ typedef void(^RLMAppNotificationBlock)(RLMApp *);
...
-+ (instancetype)uncachedAppWithId:(NSString *)appId
-                    configuration:(RLMAppConfiguration *)configuration
-                    rootDirectory:(nullable NSURL *)rootDirectory;
++ (instancetype)appWithConfiguration:(RLMAppConfiguration *)configuration;
 
 + (void)resetAppCache;
 @end
 
+@interface RLMAppConfiguration ()
+@property (nonatomic) NSString *appId;
```
Comment (dianaafanador3) [inline Realm/RLMApp_Private.h:45]: Maybe there are some user who would like to see this exposed https://github.com/realm/realm-swift/issues/8138

**Review by dianaafanador3 — APPROVED** [review]
LGTM! Just some minors comments.

File: RealmSwift/Sync.swift [hunk 2]
```diff
@@ -80,6 +80,57 @@ public typealias SyncManager = RLMSyncManager
   - see: `RLMSyncTimeoutOptions`
  */
 public typealias SyncTimeoutOptions = RLMSyncTimeoutOptions
+public extension SyncTimeoutOptions {
+    /**
+    Memberwise convenience initializer for SyncTimeoutOptions. All values are
+    in milliseconds, and use a default value if `nil`.
```
[the remark this answers is not in the document]
Comment (tgoyne) [reply RealmSwift/Sync.swift:86]: My general concern with that is there's no particular guarantee it'll be correct and these docs are very detached from where the default is actually set so it'd be very easy for it to get out of sync.

File: Realm/TestUtils/TestUtils.mm [hunk 3]
```diff
@@ -204,27 +205,24 @@ - (NSURLSession *)doStreamRequest:(RLMRequest *)request
...
+    RLMAppConfiguration *config = [[RLMAppConfiguration alloc] init];
+    config.appId = @"dummy";
+    config.encryptMetadata = false;
+    config.transport = [NoOpTransport new];
+
     // Add a fake user to the metadata Realm
     @autoreleasepool {
-        auto config = [RLMSyncManager configurationWithRootDirectory:nil appId:@"dummy"];
-        realm::SyncFileManager sfm(config.base_file_path, "dummy");
+        realm::SyncFileManager sfm(config.rootDirectory.path.UTF8String, "dummy");
```
[the remark this answers is not in the document]
Comment (tgoyne) [reply Realm/TestUtils/TestUtils.mm:215]: Yes, this is opening the metadata realm and creating a fake user for the sake of testing things which need a user but don't need to actually talk to the server.

File: Realm/RLMSyncManager.h [hunk 4]
```diff
@@ -218,11 +218,6 @@ __attribute__((deprecated("Use `RLMLogger.default`/`Logger.shared` to set/get th
 ///
 /// Defaults to 1 minute.
 @property (nonatomic) NSUInteger fastReconnectLimit;
-
-/// The app configuration that has initialized this SyncManager.
```
[the remark this answers is not in the document]
Comment (tgoyne) [reply Realm/RLMSyncManager.h:222]: This property was added to the wrong class and has never actually had a non-nil value, so it's pretty safe to assume no one is relying on it.

**Pull request merged**